Explain warning-related test failures in README (#3060) (#3102)
DataBiosphere/azul: README.md

`AZUL_DEPLOYMENT_INCARNATION` in the deployment's `environment.py` file, then
redeploy.

## Unexpected warnings cause tests to fail in `tearDownClass`

If a failure occurs due to an unexpected warning in a `tearDownClass`, add the 
warning to the list of permitted warnings found in
[`AzulTestCase`](test/azul_test_case.py) and commit the modifications. 

Due to how unit tests are discovered it's possible that a warning is emitted
when a module is loaded outside the scope of the `AzulTestCase` context manager.
The warnings that are emitted and caught are dependent on the order in which 
modules are loaded. For example, if there was an unpermitted warning emitted in a 
test case located at `test/service/test_app_logging.TestServiceAppLogging`. It's 
possible that running 
`(cd test && python -m unittest service.test_app_logging.TestServiceAppLogging)`
will not emit the warning within the scope of the `AzulTestCase` context 
manager, but instead emit the warning when the tests are loaded.
When running a test case from the directory that contains it, 
`(cd test/service && python -m unittest test_app_logging.TestServiceAppLogging)`

> the context manager might catch the warning and fail appropriately. When running

Reviewer: "might" feels like it could use more explanation.

> the test from a different directory, the test loader for some reason traverses 

Reviewer: "for some reason" could also maybe use more investigation, but if Hannes doesn't mind, it's fine with me :shrug:.

> the package structure, importing modules as it goes. This loading process might emit

Reviewer: Consider s/might/can, I think it sounds a little more definitive.

`AZUL_DEPLOYMENT_INCARNATION` in the deployment's `environment.py` file, then
redeploy.

> ## Caught warnings during tearDownClass

Reviewer 2: ```suggestion
## Unexpected warnings cause tests to fail in `tearDownClass`
```

If a failure occurs due to an unexpected warning in a `tearDownClass`, add the 
warning to the list of permitted warnings found in
[`AzulTestCase`](test/azul_test_case.py) and commit the modifications. 

Due to how unit tests are discovered it's possible that a warning is emitted
when a module is loaded outside the scope of the `AzulTestCase` context manager.
The warnings that are emitted and caught are dependent on the order in which 

> modules are loaded. For example if there was an unpermitted warning emitted in a 

Reviewer 2: ```suggestion
modules are loaded. For example, if there was an unpermitted warning emitted in a 
```

> testcase located at `test/service/test_app_logging.TestServiceAppLogging`. It's 

Reviewer 2: ```suggestion
test case located at `test/service/test_app_logging.TestServiceAppLogging`. It's 
```

Reviewer 2: Grammar. This is an incomplete sentence.

possible that running 
`(cd test && python -m unittest service.test_app_logging.TestServiceAppLogging)`
will not emit the warning within the scope of the `AzulTestCase` context 

> manager, but instead emit the warning when the tests are loaded by the 
> unit test discovery process. 

Reviewer 2: ```suggestion
manager, but instead emit the warning when the tests are loaded. 
```

> By moving in to the directory that contains the test case file, 

Reviewer 2: ```suggestion
When running a test case from the directory that contains it, 
```

`(cd test/service && python -m unittest test_app_logging.TestServiceAppLogging)`
the context manager might catch the warning and fail appropriately. When running

> the test from a different directory, `TestLoader` must traverse the directory

Reviewer 2: ```suggestion
the test from a different directory, the test loader for some reason traverses the package
```

structure, importing modules as it goes. This loading process might emit

> warnings that we otherwise would expect the context manager to catch. PyCharm's 

Reviewer 2: ```suggestion
warnings that we'd otherwise expect the context manager to catch. PyCharm's 
```

> test runner however, loads tests in a very specific order, which makes 
> `AzulTestCase` context manager more likely to catch the warning.  Within the 

Reviewer 2: ```suggestion
test runner, however, loads tests differently, which makes it more likely that the
`AzulTestCase` context manager catches the warning.  Within the 
```

continuous integration system, unit tests are discovered from within the `/test`

> directory, which causes warnings to be raised outside a context manager.

Reviewer 2: Remove this last sentence.

`AZUL_DEPLOYMENT_INCARNATION` in the deployment's `environment.py` file, then
redeploy.

## Caught warnings during tearDownClass

When running a testcase independently in PyCharm, it's possible that the 
`tearDownClass` fails due to an unexpected warning. If this failure occurs, add
the warning to the list of permitted warnings found in

> [`AzulTestCase`](test/azul_test_case.py).

Author: Ah! I can see how this can be misinterpreted, we want to commit the warnings.

Due to how unit tests are discovered it's possible that a warning is emitted
when a module is loaded outside the scope of the `AzulTestCase` context manager.
The warnings that are emitted and caught are dependent on the order in which 
modules are loaded. For example if there was an unpermitted warning emitted in a 
testcase located at `test/service/test_app_logging.TestServiceAppLogging`. It's 
possible that running 
`(cd test && python -m unittest service.test_app_logging.TestServiceAppLogging)`
will not emit the warning within the scope of the `AzulTestCase` context 
manager, but instead emit the warning when the unit tests are loaded by the 
unit test discovery process. 
By moving in to the directory that contains the test case file, 
`(cd test/service && python -m unittest test_app_logging.TestServiceAppLogging)`

> the context manager might catch the warning and fail appropriately. This is 
> because as the part of the traversal down in to the directory the discovery 
> process attempts to load modules that it finds, this loading process might emit
> warnings that we execpted the context manager to catch. When running tests for 
> a class within PyCharm, the TestRunner attempts to be very specific in how the 
> tests are loaded, which means it's more likely that the `AzulTestCase` context 
> manager catches the warning. Within the continuous integration system unit tests

Reviewer: You don't have to take all this verbatim, but I started rephrasing a sentence I found awkward and found a couple typos etc. on the way, so look closely for those other details, I don't think Github's dif is making them obvious.

```suggestion
the context manager might catch the warning and fail appropriately. When
running the test from a different directory, Python must traverse the directory
structure, importing modules as it goes. This loading process might emit
warnings that we otherwise would expect the context manager to catch. PyCharm's test
runner however, loads tests in a very specific order, which makes `AzulTestCase` 
context manager more likely to catch the warning.  Within the continuous integration system unit tests
```

Author: I added in your changes verbatim, although I'm sure Hannes will have additional input to this.

`AZUL_DEPLOYMENT_INCARNATION` in the deployment's `environment.py` file, then
redeploy.

## Unexpected warnings cause tests to fail in `tearDownClass`

Unexpected warnings that occur during testing will cause failures in 
`AzulTestCase.tearDownClass`. There is a context manager in `AzulTestCase` that
keeps record of emitted warnings during test execution. Due to the unit test 
discovery process loading modules as it traverses directories, it’s possible 
that a warning is emitted outside the scope of the context manager.

In the two commands below, the unit test discovery process occurs within a 
different directory.
```
$ (cd test && python -m unittest service.test_app_logging.TestServiceAppLogging)

> In the first case, it's possible that an unpermitted warning is emitted outside 

Reviewer 2: Why is this part of the preformatted block?

the `AzulTestCase` context manager, due to modules being loaded recursively from
the directory `test/`. If a warning is emitted outside the context manager no 

> failure will occur.

Reviewer 2: ```suggestion
test failure will occur.
```

```

```
$ (cd test/service && python -m unittest test_app_logging.TestServiceAppLogging)

> In the second case, the discovery processes loads less modules due to the 

Reviewer 2: ```suggestion
In the second case, the test discovery process loads fewer modules due to the 
```

> working directory being deeper in the `test/` directory. This may emit a warning

Reviewer 2: ```suggestion
narrowed working directory. This may emit a warning
```

during test execution, enabling the context manager to catch the unpermitted 
warning, and fail appropriately.
```

> When running tests in PyCharm the test discovery process is specific in how it 
> loads modules, this increases the chance of the `AzulTestCase` context manager
> causing a failure. If these failures occur, add the warning to the list of 
> permitted warnings found in [`AzulTestCase`](test/azul_test_case.py) and commit 
> the modifications. 

Reviewer 2: ```suggestion
Similarly, when running tests in PyCharm, its own proprietary test discovery process may also increases the chance of the `AzulTestCase` context manager
causing a failure. 

If these failures occur, add the warning to the list of 
permitted warnings found in [`AzulTestCase`](test/azul_test_case.py) and commit 
the modifications. 
```

`AZUL_DEPLOYMENT_INCARNATION` in the deployment's `environment.py` file, then
redeploy.

## Unexpected warnings cause tests to fail in `tearDownClass`

Unexpected warnings that occur during testing will cause failures in 
`AzulTestCase.tearDownClass`. There is a context manager in `AzulTestCase` that
keeps record of emitted warnings during test execution. Due to the unit test 
discovery process loading modules as it traverses directories, it’s possible 
that a warning is emitted outside the scope of the context manager.

In the two commands below, the unit test discovery process occurs within a 
different directory.

> ```

Reviewer 2: One empty line in between paragraphs.

$ (cd test && python -m unittest service.test_app_logging.TestServiceAppLogging)
```
In the first case, it's possible that an unpermitted warning is emitted outside 
the `AzulTestCase` context manager, due to modules being loaded recursively from
the directory `test/`. If a warning is emitted outside the context manager no 
test failure will occur.

> 

Reviewer 2: One empty line in between paragraphs.

```
$ (cd test/service && python -m unittest test_app_logging.TestServiceAppLogging)
```
In the second case, the test discovery process loads fewer modules due to the  
narrowed working directory. This may emit a warning during test execution, 
enabling the context manager to catch the unpermitted warning, and fail 
appropriately.

Similarly, when running tests in PyCharm, its own proprietary test discovery 
process may also increase the chance of the `AzulTestCase` context manager
causing a failure.

If these failures occur, add the warning to the list of 

> permitted warnings found in [`AzulTestCase`](test/azul_test_case.py) and commit 

Reviewer 2: W

`AZUL_DEPLOYMENT_INCARNATION` in the deployment's `environment.py` file, then
redeploy.

> ## Unexpected warnings cause tests to fail in `tearDownClass`

Reviewer 2: Two empty lines before headings.

that a warning is emitted outside the scope of the context manager.

In the two commands below, the unit test discovery process occurs within a 
different directory.

```
$ (cd test && python -m unittest service.test_app_logging.TestServiceAppLogging)
```

In the first case, it's possible that an unpermitted warning is emitted outside 
the `AzulTestCase` context manager, due to modules being loaded recursively from
the directory `test/`. If a warning is emitted outside the context manager no 
test failure will occur.

```
$ (cd test/service && python -m unittest test_app_logging.TestServiceAppLogging)
```

In the second case, the test discovery process loads fewer modules due to the  
narrowed working directory. This may emit a warning during test execution, 
enabling the context manager to catch the unpermitted warning, and fail 
appropriately.

Similarly, when running tests in PyCharm, its own proprietary test discovery 
process may also increase the chance of the `AzulTestCase` context manager
causing a failure.

If these failures occur, add the warning to the list of permitted warnings found
in [`AzulTestCase`](test/azul_test_case.py) and commit the modifications. 

> # 6. Branch flow & development process

Reviewer 2: Two empty lines before headings.
